README.md
Frequency is a Polkadot parachain designed to run Decentralized Social Network Protocol (DSNP), but it could run other things.

# Table of Contents

-   [Table of Contents](#table-of-contents)
-   [Prerequisites](#prerequisites)
    -   [Hardware](#hardware)
-   [Build](#build)
    -   [Local Desktop](#local-desktop)
        -   [asdf Support](#asdf-support)
    -   [Remote Instance such as AWS EC2](#remote-instance-such-as-aws-ec2)
        -   [Ubuntu](#ubuntu)
-   [Run](#run)
    -   [1. Collator Node in Instant Sealing Mode](#1-collator-node-in-instant-sealing-mode)
        -   [In Terminal](#in-terminal)
        -   [In Docker Container](#in-docker-container)
    -   [2. Collator Node with Local Relay Chain](#2-collator-node-with-local-relay-chain)
        -   [Mixed Terminal/Docker](#mixed-terminaldocker)
            -   [Stop and Clean Environment](#stop-and-clean-environment)
        -   [All in Docker Container](#all-in-docker-container)
    -   [Run Tests](#run-tests)
    -   [Run Benchmarks](#run-benchmarks)
    -   [Generate Specs](#generate-specs)
-   [Format, Lint and Audit Source Code](#format-lint-and-audit-source-code)
-   [Verify Runtime](#verify-runtime)
-   [Local Runtime Upgrade](#local-runtime-upgrade)
-   [Contributing](#contributing)
-   [Additional Resources](#additional-resources)
-   [Miscellaneous](#miscellaneous)

# Prerequisites

1. [Docker Engine](https://docs.docker.com/engine/install/)\*
1. [Docker Compose](https://docs.docker.com/compose/install/)

---

-   For Mac users, [Docker Desktop](https://docs.docker.com/desktop/mac/install/) engine also installs docker compose environment, so no need to install it separately.

## Hardware

We run benchmarks with and recommend the same [reference hardware specified by Parity](https://wiki.polkadot.network/docs/maintain-guides-how-to-validate-polkadot#reference-hardware).

# Build

## Local Desktop

1. Install Rust using the [official instructions](https://www.rust-lang.org/tools/install).
2. Check out this repository
3. `rust-toolchain.toml` specifies the standard toolchain to use. If you have `rustup` installed, it will automatically install the correct toolchain when you run any cargo command.
4. Running `make check` will run cargo checks for all frequency features. This is the recommended way to check your code before committing. Alternatively, you can run following for specific features:

    ```sh
    make check-local
    make check-rococo
    make check-mainnet
    ```

5. Build Wasm and native code.

    _Note, if you get errors complaining about missing
    dependencies (cmake, yarn, node, jq, etc.) install them with your favorite package
    manager(e.g. Homebrew on Mac) and re-run the command again._

    ```sh
    make build
    ```

    Above will build frequency with all frequency features. Alternatively you may run following command to build with specific features:

    ```sh
    make build-local
    make build-rococo
    make build-mainnet
    ```

    To build local, rococo or mainnet features respectively.

At this point you should have `./target/release` directory generated locally with compiled
project files.

### asdf Support

Install the required plugins for asdf:

```sh
asdf plugin-add rust
asdf plugin-add make
asdf plugin-add cmake [https://github.com/srivathsanmurali/asdf-cmake.git](https://github.com/srivathsanmurali/asdf-cmake.git)
```

Install the dependency versions declared in `.tool-versions`

```sh
asdf install
```

NOTE: I could find no clang plugin that worked so your system still needs clang to be installed.

## Remote Instance such as AWS EC2

For remote instances running Linux, if you want to check out and build such as on an AWS EC2 instance, the process is slightly different to what is in the [Substrate documentation](https://docs.substrate.io/main-docs/install/linux/).

### Ubuntu

1. Upgrade the instance and install missing packages with `apt`:

```bash
sudo apt upgrade
sudo apt upgrade git
sudo apt install —-assume-yes build-essential
sudo apt install --assume-yes clang curl libssl-dev cmake
```

2. Follow [official instructions to install rust](https://www.rust-lang.org/tools/install), but select `3. customize the installation`, then reply **n** to `Modify PATH variable? (Y/n)`
3. Follow steps 6-10 at [Substrate: Linux development](https://docs.substrate.io/main-docs/install/linux/)
4. Proceed with checking out and building frequency as above.

# Run

There are 2 options to run the chain locally:

_Note, Running frequency via following options does not require binary to be built or chain specs to be generated separately, and is programmed within the scripts for simplicity._

1.  Collator Node in Instant Sealing Mode,
2.  Collator Node with Local Relay Chain

## 1. Collator Node in Instant Sealing Mode

![](docs/images/local-dev-env-option-1.jpg)

This option runs just one collator node in instant seal mode and nothing more.
A "collator node" is a Frequency parachain node that is actively collating (aka forming blocks to submit to the relay chain). The instant seal mode allows a blockchain node to author a block
as soon as it goes into a queue. This is also a great option to run with an example client.

### In Terminal

```sh
make start
```

### In Docker Container

```sh
docker run --rm -p 9944:9944 -p 9933:9933 -p 30333:30333 frequencychain/instant-seal-node
```

To stop running chain hit [Ctrl+C] in terminal where the chain was started.

| **Node**                |             **Ports**             | **Explorer URL**                                                                          |
| ----------------------- | :-------------------------------: | ----------------------------------------------------------------------------------------- |
| Frequency Collator Node | ws:`9944`, rpc`:9933`, p2p:`3033` | [127.0.0.1:9944](https://polkadot.js.org/apps/?rpc=ws%3A%2F%2F127.0.0.1%3A9944#/explorer) |

## 2. Collator Node with Local Relay Chain

![](docs/images/local-dev-env-option-2.jpg)

### Mixed Terminal/Docker

This option runs one collator node as local host process and two relay chain validator nodes in each own docker container.

1. Start relay chain validator nodes.

    ```sh
    make start-relay
    ```

1. Register a new parachain slot (parachain id) for Frequency. _Note, if parachain was
   previously registered on a running relay chain and no new registration is required,
   then you can skip the above step._
    ```sh
    make register
    ```
1. Generate chain spec files. If this is your first time running the project or
   new pallets/runtime code changes have been made to Frequency, then the chain specs
   need to be generated. Refer to [generation spec file](#generate-a-new-spec-file)
   for more details.

1. Start Frequency as parachain. This step will generate genesis/wasm and onboard the
   parachain.

    ```sh
    make start-frequency
    ```

1. Onboard Frequency to the relay chain
    ```sh
    make onboard
    ```

#### Stop and Clean Environment

1. Off-board Frequency from relay chain.: `make offboard`
2. to stop Frequency running in the terminal: `[Ctrl+C] `
3. Stop the relay chain. `make stop-relay`
4. Run to remove unused volumes. `make docker-prune`
5. Clean up temporary directory to avoid any conflicts with next onboarding:
   `rm -fr /tmp/frequency`

### All in Docker Container

:exclamation: Currently does not work on M\* series MacOS laptops. See https://github.com/LibertyDSNP/frequency/issues/432

Start:

```sh
make start-frequency-docker
```

Stop:

```sh
make stop-frequency-docker
```

| **Node**             | **Ports**                           | **Explorer URL**                                                                          |
| -------------------- | ----------------------------------- | ----------------------------------------------------------------------------------------- |
| Frequency Relay Node | ws:`9944`, rpc`:9933`, p2p:`30333`  | [127.0.0.1:9944](https://polkadot.js.org/apps/?rpc=ws%3A%2F%2F127.0.0.1%3A9944#/explorer) |
| Alice Relay Node     | ws:`:9946`, rpc`:9935`, p2p:`30335` | [127.0.0.1:9946](https://polkadot.js.org/apps/?rpc=ws%3A%2F%2F127.0.0.1%3A9946#/explorer) |
| Bob Relay Node       | ws:`:9947`, rpc`:9936`, p2p:`30336` | [127.0.0.1:9947](https://polkadot.js.org/apps/?rpc=ws%3A%2F%2F127.0.0.1%3A9947#/explorer) |

## Run Tests

```sh
# Activate selected features
cargo test --features runtime-benchmarks, std

# Activate all features and test all packages in the workspace
cargo test --all-features --workspace --release
```

## Run Benchmarks

```sh
make benchmarks
```

# Format, Lint and Audit Source Code

-   Format code with `make format` according to style guidelines and configurations in `rustfmt.toml`.
-   Lint code with with `make lint` to catch common mistakes and improve your [Rust](https://github.com/rust-lang/rust) code.
-   Alternatively, run `make format-lint` to run both at the same time.
-   Run `cargo-deny` to audit Cargo.lock files for crates with security vulnerabilities reported to the RustSec Advisory Database. [See cargo-deny installation instructions](https://github.com/EmbarkStudios/cargo-deny)

# Verify Runtime

1. Check out the commit at which the runtime was built.
2. Use srtool to verify the runtime:
    ```sh
    TARGET=build-runtime RUST_TOOLCHAIN=nightly ./tools/scripts/build.sh
    ```

# Local Runtime Upgrade

To upgrade the runtime, run the following command:

```sh
make upgrade-local
```

The current scripts follow this process for upgrading locally:

1. Build new runtime and generate the compressed wasm: `make specs-rococo-2000`
2. Call `authorizeUpgrade` extrinsic from parachain system to initiate the upgrade.
3. Call `enactAuthorizedUpgrade` extrinsic from parachain system to enact the upgrade.
4. For testnet and mainnet, the upgrade is done slightly differently using `scheduler` and enactment is scheduled for a specific block number in the future.

# Contributing

Interested in contributing?
Wonderful!
Please check out [the information here](./CONTRIBUTING.md).

# Additional Resources

-   [Cumulus Project](https://github.com/paritytech/cumulus)
-   [Cumulus Tutorials](https://docs.substrate.io/tutorials/)
-   [Prep Substrate environment for development](https://docs.substrate.io/install/)

# Miscellaneous

```sh
# Clean up local docker resources
make docker-prune

# View all listening ports
lsof -i -P | grep -i "listen"

# View ports Frequency node is listening on
lsof -i -P | grep -i "listen" | grep frequency
```


CHANGELOG.md
# Changelog

## v0.9.29-rc1 - 2022-10-22

-   Initial Release


scripts/init.sh
#!/usr/bin/env bash

set -e

cmd=$1
chain_spec="${RAW_PARACHAIN_CHAIN_SPEC:-./res/genesis/local/rococo-local-frequency-2000-raw.json}"
# The runtime we want to use
parachain="${PARA_CHAIN_CONFIG:-rococo-2000}"
# The parachain Id we want to use
para_id="${PARA_ID:-2000}"
# The tmp base directory
base_dir=/tmp/frequency
# Option to use the Docker image to export state & wasm
docker_onboard="${DOCKER_ONBOARD:-false}"
frequency_docker_image_tag="${PARA_DOCKER_IMAGE_TAG:-frequency-latest}"
chain="${RELAY_CHAIN_SPEC:-./resources/rococo-local.json}"

case $cmd in

start-relay-chain)
  echo "Starting local relay chain with Alice and Bob..."
  cd docker
  docker-compose up -d relay_alice relay_bob
  ;;

stop-relay-chain)
  echo "Stopping relay chain..."
  cd docker
  docker-compose down
  ;;

start-frequency-docker)
  echo "Starting frequency container with Alice..."
  cd docker
  docker-compose up --build collator_frequency
  ;;

stop-frequency-docker)
  echo "Stopping frequency container with Alice..."
  cd docker
  docker-compose down
  ;;

start-frequency)
  printf "\nBuilding frequency with runtime '$parachain' and id '$para_id'...\n"
  cargo build --release --features frequency-rococo-local

  parachain_dir=$base_dir/parachain/${para_id}
  mkdir -p $parachain_dir;

  if [ "$2" == "purge" ]; then
    echo "purging parachain..."
    rm -rf $parachain_dir
  fi

  ./scripts/run_collator.sh \
    --chain="frequency-local" --alice \
    --base-path=$parachain_dir/data \
    --wasm-execution=compiled \
    --execution=wasm \
    --force-authoring \
    --port $((30333)) \
    --rpc-port $((9933)) \
    --ws-port $((9944)) \
    --rpc-external \
    --rpc-cors all \
    --ws-external \
    --rpc-methods=Unsafe \
    --state-cache-size 0 \
  ;;

start-frequency-instant)
  printf "\nBuilding frequency with runtime instant sealing ...\n"
  cargo build --release --features frequency-rococo-local

  parachain_dir=$base_dir/parachain/${para_id}
  mkdir -p $parachain_dir;

  if [ "$2" == "purge" ]; then
    echo "purging parachain..."
    rm -rf $parachain_dir
  fi

  ./target/release/frequency \
    --dev \
    -lruntime=debug \
    --instant-sealing \
    --wasm-execution=compiled \
    --execution=wasm \
    --no-telemetry \
    --no-prometheus \
    --port $((30333)) \
    --rpc-port $((9933)) \
    --ws-port $((9944)) \
    --rpc-external \
    --rpc-cors all \
    --ws-external \
    --rpc-methods=Unsafe \
    --tmp
  ;;

start-frequency-container)

  parachain_dir=$base_dir/parachain/${para_id}
  mkdir -p $parachain_dir;
  frequency_default_port=$((30333))
  frequency_default_rpc_port=$((9933))
  frequency_default_ws_port=$((9944))
  frequency_port="${Frequency_PORT:-$frequency_default_port}"
  frequency_rpc_port="${Frequency_RPC_PORT:-$frequency_default_rpc_port}"
  frequency_ws_port="${Frequency_WS_PORT:-$frequency_default_ws_port}"

  ./scripts/run_collator.sh \
    --chain="frequency-local" --alice \
    --base-path=$parachain_dir/data \
    --wasm-execution=compiled \
    --execution=wasm \
    --force-authoring \
    --port "${frequency_port}" \
    --rpc-port "${frequency_rpc_port}" \
    --ws-port "${frequency_ws_port}" \
    --rpc-external \
    --rpc-cors all \
    --ws-external \
    --rpc-methods=Unsafe \
    --state-cache-size 0 \
  ;;

register-frequency)
  echo "reserving and registering parachain with relay via first available slot..."

  cd scripts/js/onboard
  yarn && yarn register "ws://0.0.0.0:9946" "//Alice"
  ;;

onboard-frequency)
  echo "Onboarding parachain with runtime '$parachain' and id '$para_id'..."

   onboard_dir="$base_dir/onboard"
   mkdir -p $onboard_dir

   wasm_location="$onboard_dir/${parachain}-${para_id}.wasm"
    if [ "$docker_onboard" == "true" ]; then
      genesis=$(docker run -it {REPO_NAME}/frequency:${frequency_docker_image_tag} export-genesis-state --chain="frequency-local")
      docker run -it {REPO_NAME}/frequency:${frequency_docker_image_tag} export-genesis-wasm --chain="frequency-local" > $wasm_location
    else
      genesis=$(./target/release/frequency export-genesis-state --chain="frequency-local")
      ./target/release/frequency export-genesis-wasm --chain="frequency-local" > $wasm_location
    fi

  echo "WASM path:" "${parachain}-${para_id}.wasm"

  cd scripts/js/onboard
  yarn && yarn onboard "ws://0.0.0.0:9946" "//Alice" ${para_id} "${genesis}" $wasm_location
  ;;

offboard-frequency)
  echo "cleaning up parachain for id '$para_id'..."

  cd scripts/js/onboard
  yarn && yarn cleanup "ws://0.0.0.0:9946" "//Alice" ${para_id}
  ;;

upgrade-frequency)

  root_dir=$(git rev-parse --show-toplevel)

  cargo build \
    --locked \
    --profile release \
    --package frequency-rococo-runtime \
    --target-dir $root_dir/target/upgrade \
    -Z unstable-options


  wasm_location=$root_dir/target/upgrade/release/wbuild/frequency-rococo-runtime/frequency_rococo_runtime.compact.compressed.wasm

  ./scripts/runtime-upgrade.sh "//Alice" "ws://0.0.0.0:9944" $wasm_location

  ./scripts/enact-upgrade.sh "//Alice" "ws://0.0.0.0:9944" $wasm_location

  ;;

esac
